sys/net/gnrc/link_layer/csma_sender/gnrc_csma_sender.c

```diff
@@ -0,0 +1,245 @@
+                               sizeof(netopt_enable_t));
+    bool ok = false;
+    switch (res) {
+    case -ENODEV:
+        /* invalid device pointer given */
+        return -ENODEV;
+    case -ENOTSUP:
+        /* device doesn't make auto-CSMA/CA */
+        ok = false;
+        break;
+    case -EOVERFLOW:  /* (normally impossible...*/
+    case -ECANCELED:
+        DEBUG("!!! DEVICE DRIVER FAILURE! TRANSMISSION ABORTED!\n");
+        /* internal driver error! */
+        return -ECANCELED;
+    default:
+        ok = (hwfeat == NETOPT_ENABLE);
+    }
+
+    if (ok) {
+        /* device does CSMA/CA all by itself: let it do its job */
+        DEBUG("Network device does hardware CSMA/CA\n");
+        return dev->driver->send_data(dev, pktsnip);
+    }
+
+    /* if we arrive here, then we must perform the CSMA/CA procedure
+       ourselves by software */
+    genrand_init(xtimer_now());
+    DEBUG("Starting software CSMA/CA....\n");
+
+    int nb = 0, be = macMinBE;
+
+    while (nb <= macMaxCSMABackoffs) {
+        /* delay for an adequate random backoff period */
+        uint32_t bp = choose_backoff_period(be);
+        xtimer_usleep(bp);
+
+        /* try to send after a CCA */
+        res = send_if_cca(dev, pkt);
+        if (res >= 0) {
+            /* TX done */
+            return res;
+        } else if (res != -EBUSY) {
+            /* something has gone wrong, return the error code */
+            return res;
+        }
+
+        /* medium is busy: increment CSMA counters */
+        DEBUG("Radio medium busy.\n");
+        be++;
+        if (be > macMaxBE) {
+            be = macMaxBE;
+        }
+        nb++;
+        /* ... and try again if we have no exceeded the retry limit */
+    }
+
+    /* if we arrive here, medium was never available for transmission */
+    DEBUG("Software CSMA/CA failure: medium never available.\n");
+    return -EBUSY;
+}
+
+
+int cca_send(gnrc_netdev_t *dev, gnrc_pktsnip_t *pkt)
+{
+    netopt_enable_t hwfeat;
+
+    /* Does the transceiver do automatic CCA before sending? */
+    int res = dev->driver->get(dev,
+                               NETOPT_AUTOCCA,
+                               (void *) &hwfeat,
+                               sizeof(netopt_enable_t));
+    bool ok = false;
+    switch (res) {
+    case -ENODEV:
+        /* invalid device pointer given */
+        return -ENODEV;
+    case -ENOTSUP:
+        /* device doesn't make auto-CCA */
+        ok = false;
+        break;
+    case -EOVERFLOW:  /* (normally impossible...*/
+    case -ECANCELED:
+        /* internal driver error! */
+        DEBUG("!!! DEVICE DRIVER FAILURE! TRANSMISSION ABORTED!\n");
+        return -ECANCELED;
+    default:
+        ok = (hwfeat == NETOPT_ENABLE);
+    }
+
+    if (ok) {
+        /* device does auto-CCA: let him do its job */
+        DEBUG("Network device does auto-CCA checking.\n");
+        return dev->driver->send_data(dev, pktsnip);
+    }
```

Comment:
In my version, I can also print a debug message when the hardware's auto-CCA feature is actually used. It could prove useful in debugging sessions with devices that would somehow have a problem with this automatic CCA checking (I think especially to AT86RF231 in "extended modes", as @daniel-k explained above; maybe other transceivers may show similar problems in the future).